FlowCrypt/Functionality/Services/Compose Message Service/ComposeMessageService.swift

```diff
@@ -0,0 +1,137 @@
+        self.dataService = dataService
+        self.contactsService = contactsService
+        self.core = core
+    }
+
+    // MARK: - Validation
+    func validateMessage(
+        input: ComposeMessageInput,
+        contextToSend: ComposeMessageContext,
+        email: String,
+        atts: [SendableMsg.Attachment]
+    ) -> Result<SendableMsg, ComposeMessageError> {
+        let recipients = contextToSend.recipients
+
+        guard recipients.isNotEmpty else {
+            return .failure(.validationError(.emptyRecipient))
+        }
+
+        let emails = recipients.map(\.email)
+        let hasContent = emails.filter { $0.hasContent }
+
+        guard emails.isNotEmpty else {
+            return .failure(.validationError(.emptyRecipient))
+        }
+
+        guard emails.count == hasContent.count else {
+            return .failure(.validationError(.emptyRecipient))
+        }
+
+        guard input.isReply || contextToSend.subject?.hasContent ?? false else {
+            return .failure(.validationError(.emptySubject))
+        }
+
+        guard let text = contextToSend.message, text.hasContent else {
+            return .failure(.validationError(.emptyMessage))
+        }
+
+        let subject = input.subjectReplyTitle
+            ?? contextToSend.subject
+            ?? "(no subject)"
+
+        guard let myPubKey = self.dataService.publicKey() else {
+            return .failure(.validationError(.missedPublicKey))
+        }
+
+        switch getPubKeys(for: recipients) {
+        case .success(let allRecipientPubs):
+            let replyToMimeMsg = input.replyToMime
+                .flatMap { String(data: $0, encoding: .utf8) }
+
+            let msg = SendableMsg(
+                text: text,
+                to: recipients.map(\.email),
+                cc: [],
+                bcc: [],
+                from: email,
+                subject: subject,
+                replyToMimeMsg: replyToMimeMsg,
+                atts: atts,
+                pubKeys: allRecipientPubs + [myPubKey]
+            )
+
+            return .success(msg)
+        case .failure(let error):
+            return .failure(.validationError(error))
+        }
+    }
+
+    private func getPubKeys(for recepients: [ComposeMessageRecipient]) -> Result<[String], MessageValidationError> {
+        let pubKeys = recepients.map {
+            ($0.email, contactsService.retrievePubKey(for: $0.email))
+        }
+
+        let emailsWithoutPubKeys = pubKeys.filter { $0.1 == nil }.map(\.0)
+
+        guard emailsWithoutPubKeys.isEmpty else {
+            return .failure(.noPubRecipients(emailsWithoutPubKeys))
+        }
+
+        return .success(pubKeys.compactMap(\.1))
+    }
+
+    // MARK: - Encrypt and Send
+    func encryptAndSend(message: SendableMsg) -> AnyPublisher<Void, ComposeMessageError> {
+        messageGateway.sendMail(mime: encryptMessage(with: message))
+            .mapError { ComposeMessageError.gatewayError($0) }
```

Comment:
What does `.mapError` with `.eraseToAnyPublisher` do?

Reply:
`func sendMail(mime: Data) -> Future<Void, Error>` will return just Error for us. But we are expecting to have `ComposeMessageError`, so mapError in case of error will transform this to `ComposeMessageError.gatewayError` with associated error. either from Imap or Gmail api. 

`eraseToAnyPublisher` - exposes an instance of AnyPublisher to the downstream subscriber, rather than this publisher’s actual type. 
So without this it will have a return type as `Publishers.MapError<Future<Void, Error>, ComposeMessageError>`. 
Please don't ask why)) Swift - 🤷
Type erasure is not the best in swift nowadays. 
https://developer.apple.com/documentation/combine/just/erasetoanypublisher()